### PR TITLE
ECS vs EKS blog post

### PR DESCRIPTION
**Pull request by logandonley**
My brain is a bit fried right now, so I feel like I've missed some topics I had intended to cover. 

If you get a chance, I'd appreciate a review of what I've got and any suggestions you may have.

### CODE REVIEW

File: content/posts/eks-vs-ecs.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+title: Amazon ECS vs EKS for running CloudBees Core/Jenkins
+authors:
+  - "Logan Donley"
+date: 2020-01-13T05:05:15-04:00
+showDate: true
+tags: ["Kubernetes","aws","EKS","ECS"]
+draft: true
+---
+
+A question my colleagues and I are often asked is "if I'm already using ECS is there any point in moving to EKS (or Kubernetes in general) for running CloudBees Core?" Both Amazon ECS and EKS are great products, but we have a strong preference towards Kubernetes. We'll take a look at why that is in this post.
```
Comment (kmadel) [inline content/posts/eks-vs-ecs.md]: Instead of "We'll take a look at why that is in this post." Perhaps just "We'll take a look at why in this post." Less wordy is usually better.

File: content/posts/eks-vs-ecs.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+These are both services from AWS which deal with container orchestration. Containers have become a crucial component in the modern tech stack, but as your applications scale, having an orchestrator is a must. 
+
+### ECS
+
+Amazon Elastic Container Service is a platform which will orchestrate your containers in a straightforward manner. You create an application definition and it will ensure that the container is running. 
+
+With ECS there are a couple of ways you can run your containers: on EC2 instances or with [Fargate](https://aws.amazon.com/fargate/), AWS's serverless compute-engine.
+
+With EC2 you operate like you normally would and configure auto-scaling groups, networking, etc. Fargate on the other hand abstracts away the individual server and has you only worry about cpu and memory needs. The cost will be higher per compute/memory resource but it handles a lot for you.
```
Comment (kmadel) [inline content/posts/eks-vs-ecs.md]: "has you only worry about cpu and memory needs" - slightly strange phrasing. Also, is that all you have to 'worry' about with Fargate?

Finally, not clear you are referring to Fargate when discussing cost.

File: content/posts/eks-vs-ecs.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+
+### EKS
+
+Amazon Elastic Kubernetes Service is AWS's managed Kubernetes platform. Kubernetes has become the community-favored container orchestrator and for good reason. It's incredibly powerful, resilient, and now has a great ecosystem built around it.
+
+With all of the power of Kubernetes comes some complexity on the infrastructure side. Setting up and maintaining your own Kubernetes cluster can be challenging even for those with lots of experience. That's why whenever possible, we would recommend you use a managed Kubernetes service like EKS, GKE, or AKS since they solve those challenges for you.
```
Comment (kmadel) [inline content/posts/eks-vs-ecs.md]: Comma after challenging? "... challenging, even for those with lots of experience."

File: content/posts/eks-vs-ecs.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+Using EKS is in many ways similar to using ECS since whether you use Fargate or EC2, AWS will take care of provisioning the resources for you, letting you instead focus on the applications you will be running. The real differences come down to which container orchestrator you are using. ECS's solution or Kubernetes'.
+
+## Battle of the orchestrators
+
+If at the end of the day an application is going to be running in a container does it really matter which orchestrator you use as long as it works? Especially if one is simpler than the other? I would argue that it does matter. 
+
+Container orchestrators have become the new platform on which you build and deploy applications. While we often talk about vendor lock-in as a situation many people get into, it happens just as frequently with technologies. For instance, if you are developing for Linux machines your application isn't necessarily portable to a Windows machine. At least not without extra work. This is a form of lock-in.
```
Comment (kmadel) [inline content/posts/eks-vs-ecs.md]: "about vendor lock-in as a situation many people get into" to "about avoiding vendor lock-in as a best practice"

"it happens just as frequently with technologies" - "with other technologies."?

File: content/posts/eks-vs-ecs.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+
+| Cons |
+| --- |
+| - Kubernetes can be challenging to learn |
+| - Without automation or tooling, the process of deploying a single application is more involved |
+| - Can be overkill for small projects |
+
+
+* It's important to note that with the Fargate integration, there are some limitations such as which regions you can deploy to [https://docs.aws.amazon.com/eks/latest/userguide/fargate.html](https://docs.aws.amazon.com/eks/latest/userguide/fargate.html).
```
Comment (kmadel) [inline content/posts/eks-vs-ecs.md]: Also, I thought that Fargate on EKS doesn't support stateful applications yet - so that is a drawback for specific types of applications - like Jenkins. Or has that changed?

File: content/posts/eks-vs-ecs.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+
+## Running CloudBees Core/Jenkins
+
+Now the chances are you are reading this to learn about which platform to use for CloudBees Core or Jenkins. As mentioned earlier, we have a preference for Kubernetes for a few of reasons which I will get into here.
```
Comment (kmadel) [inline content/posts/eks-vs-ecs.md]: Suggest rewriting to affirm the purpose of this post - make it about providing an opinion on what to use for Core vs letting the reader decide if that is why they are reading this.

File: content/posts/eks-vs-ecs.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+### Portability
+
+Kubernetes is an open-standard. When you build your CI/CD system on Kubernetes, it can run on EKS, GKE, AKS, OpenShift, PKS, home-built, etc. without issue (well some vendor flavors like OpenShift have additional resources which aren't portable to standard k8s clusters).
+
+Some organizations are fine with going all-in on a single cloud vendor, but many others try to avoid vendor lock-in as much as possible. Keeping to as many open and portable platforms as possible ensures that the effort you do while building your system won't be wasted if you switch from AWS to GCP or Azure.
```
Comment (kmadel) [inline content/posts/eks-vs-ecs.md]: Maybe "Using" or "Leveraging" instead of "Keeping"?

File: content/posts/eks-vs-ecs.md [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+The ECS platform exists only in AWS and is proprietary, so you don't have portability to other vendors. 
+
+This is a big plus for using EKS (or any Kubernetes).
+
+
+### Community
+
+As of writing this post, Kubernetes is the open standard for container orchestration, in major part because of it's huge and rapid community adoption. Since it is the standard, a large amount of DevOps tooling has been built on top of it. Most CI/CD processes will run through many different tools, so opting for the platform with the greatest marketshare is a safe choice.
+
+Because of this being the reality, we at CloudBees have built lots of functionality into CloudBees Core specific to Kubernetes which adds a lot of quality of life improvements. The ability to dynamically provision masters and agents is in my opinion a game changer. There are features like hibernating masters which we have recently released as a technical preview which allows for potentially huge cost savings by spinning down masters that haven't been used in a while. Functionality like this is only possible on a consistent platform, and Kubernetes gives us this.
```
Comment (kmadel) [inline content/posts/eks-vs-ecs.md]: "Because of this being the reality" to "Because of this reality"

Also may be nice to throw in something around how Core on K8s provides a platform for easily running and managing many smaller Jenkins Masters.

File: content/posts/eks-vs-ecs.md [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+
+## Wrapping it up
+
+When it comes to running CloudBees Core on a modern platform, Kubernetes is our recommendation to you if it is at all possible. With recent changes to Amazon's EKS to support Fargate, it is becoming even easier to manage your environment.
```
Comment (kmadel) [inline content/posts/eks-vs-ecs.md]: I would remove "if it is at all possible" - anything is possible and we should be more firm with our opinions. 

Mentioning Fargate on EKS seems a bit misleading - at least in the context of Core or Jenkins since I don't believe it is possible to run either there yet.

File: content/posts/eks-vs-ecs.md [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+ECS is a nice tool, especially for smaller projects, but it pales in comparison to Kubernetes when it comes to the qualities which the DevOps community tend to look for.
```
Comment (kmadel) [inline content/posts/eks-vs-ecs.md]: would reword "qualities which the DevOps community tend to look for" - something like:

the quality features required by the DevOps community.

**Comment by logandonley** [comment]
Thanks for the review, I've updated with all of those suggestions.

**Comment by kmadel** [comment]
@logandonley  Looks good, let us know when you are ready for it to be published.
